Ht 2503 dataden

moseshll opened the pull request:
More work in progress. Zip and METS check (`sub check_zip()`) should perhaps be removed or made optional depending on how pedantic we wish to be.

aelkiss committed:
HT-2503: WIP for data den auditing

moseshll committed:
Add more checks to Data Den audit.

aelkiss commented on audit_data_den.pl:
```diff
@@ -0,0 +1,267 @@
+my $base = shift @ARGV or die("Missing base directory..");
+$base .= '/' unless substr($base, -1, 1) eq '/';
+
+my $last_run = '0000-00-00 00:00:00';
```
Wondering if it might be easier to just record when the program started and at the end look for items that haven't been updated since the start of the run?

moseshll replied:
To clarify, you recommend swapping out the `select max(lastchecked) from feed_backups` altogether? If so then I'd probably just swap in `select now()`. (I'm leery of mixing MariaDB and system times because of the potential for time zone madness.)

aelkiss replied:
Good point. I think `select now()` makes sense.

aelkiss commented on audit_data_den.pl:
```diff
@@ -0,0 +1,267 @@
+while ( my $line = <RUN> ) {
+  chomp($line);
+
+  #next if $line =~ /\Qpre_uplift.mets.xml\E/;
```
We can remove this line.

aelkiss commented on audit_data_den.pl:
```diff
@@ -0,0 +1,267 @@
+  }
+}
+
+sub check_zip {
```
Could we move this to a separate script that accepts just a single path? For time time being we don't want to ever accidentally do a bulk read of items under data den (since it might trigger large numbers of recalls of items from tape)

moseshll replied:
Spun off in next PR.

aelkiss reviewed (changes requested):
Looks good except zipcheck needs to be excised so we don't accidentally recall large numbers of items from tape. Once that's done I can give it a try on the real thing and see if there any issues.

moseshll committed:
Spin zipfile check off to new scripts

moseshll commented:
I think this may be ready to unleash on real data with the new commit.

aelkiss commented on audit_data_den_zipfile.pl:
```diff
@@ -0,0 +1,117 @@
+$zipfile = Cwd::abs_path($zipfile);
+die("zip file $zipfile does not exist") unless -f $zipfile;
+my $pathinfo = DataDenPath->new($zipfile);
+check_zip($pathinfo);
```
we should update feed_audit and set lastmd5check = NOW(), md5check_ok = 1 if it succeeded and lastmd5check = NOW(), md5check = 0 if it failed.

moseshll replied:
I'm reading `ht_repository.feed_audit` as `ht_repository.feed_backups`, lemme know if that is not OK.

aelkiss commented on run_data_den_audit.sh:
```diff
@@ -0,0 +1,8 @@
+
+root="${1:-/htdataden/}"
+
+export HTFEED_CONFIG=$SDRROOT/feed/etc/config_dev.yaml
```
Let's set `HTFEED_CONFIG` and `FEED_HOME` only if they aren't already set. One way to do this is (allegedly):

```bash
: ${HTFEED_CONFIG:=$SDRROOT/feed/etc/config_dev.yaml}
: ${FEED_HOME:=$SDRROOT/feed}
```

aelkiss commented on audit_data_den.pl:
```diff
@@ -0,0 +1,196 @@
+my $select_now =
+'select now()';
+
+my $select_unchecked =
```
Probably worth at least extracting this part as a function. I'm waffling about how to actually run this in production; this might make more sense as a more explicitly structured separately-run step, which would also make it easier to try to parallelize the first part of the checking (just run several copies at different parts of the pairtree)

aelkiss committed:
HT-2503: Data Den audit cleanup

- extract function for looking for missing items in data den so we can
have an option for that later, if we want to parallelize the
looking-through-data-den part

- update feed_backups table when doing md5 check

- Only set FEED_HOME and HTFEED_CONFIG if they aren't already set
